File: strategy_engine.py
import logging
import time
import math

logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s] %(message)s")
logger = logging.getLogger("Strategy_Engine")

class OKXGridValidator:
    def __init__(self, min_order_val: float = 1.0, min_grids: int = 2, max_grids: int = 150):
        self.min_order_val = min_order_val
        self.min_grids = min_grids
        self.max_grids = max_grids

    def validate_setup(self, lower_price: float, upper_price: float, grid_count: int, 
                       total_investment: float, current_market_price: float) -> tuple:
        
        # 1. Range bounds check
        if lower_price >= upper_price:
            return False, "Error: Lower price must be strictly less than upper price."
        
        # 2. Market price containment check (OKX requirement for Spot Grid)
        if not (lower_price <= current_market_price <= upper_price):
            return False, f"Error: Price (${current_market_price}) is outside grid range [${lower_price}, ${upper_price}]."

        # 3. Grid count boundaries
        if not (self.min_grids <= grid_count <= self.max_grids):
            return False, f"Error: Grid count must be between {self.min_grids} and {self.max_grids}."

        # 4. Investment per grid (OKX minimum sizing check)
        investment_per_grid = total_investment / grid_count
        if investment_per_grid < self.min_order_val:
            return False, (
                f"Error: Investment per grid (${investment_per_grid:.2f}) is below "
                f"OKX minimum threshold (${self.min_order_val:.2f})."
            )

        return True, "Validation Passed"

class GridStrategyEngine:
    def __init__(self, inst_id: str, lower_bound: float, upper_bound: float, grids: int, investment_amount: float, grid_type: str = "arithmetic"):
        self.inst_id = inst_id
        self.lower_bound = lower_bound
        self.upper_bound = upper_bound
        self.grids = grids
        self.investment_amount = investment_amount
        self.grid_type = grid_type.lower()
        self.grid_levels = []
        self.last_grid_index = None
        self.active = False
        self.realized_pnl = 0.0
        self.current_pos = 0.0
        self.avg_price = 0.0
        self._initialize_grid()

    def _initialize_grid(self):
        """Calculates the price points for each grid buy/sell level."""
        if self.grids <= 1:
            raise ValueError("Grid count must be greater than 1.")

        if self.grid_type == "geometric":
            # Geometric ratio
            ratio = (self.upper_bound / self.lower_bound) ** (1 / (self.grids - 1))
            self.grid_levels = [round(self.lower_bound * (ratio ** i), 2) for i in range(self.grids)]
        else:
            # Arithmetic (default)
            step = (self.upper_bound - self.lower_bound) / (self.grids - 1)
            self.grid_levels = [round(self.lower_bound + (i * step), 2) for i in range(self.grids)]
            
        self.grid_levels = sorted(self.grid_levels)
        logger.info(f"Initialized {self.grid_type.upper()} Grid Strategy for {self.inst_id} with {self.grids} levels: {self.grid_levels}")

    def process_tick(self, current_price: float):
        """
        Evaluates the current market price and returns a signal if a grid level is crossed.
        Returns: (signal_type, price, size) or None
        """
        if not self.active:
            # Initial anchor point
            for i, level in enumerate(self.grid_levels):
                if current_price >= level:
                    self.last_grid_index = i
            self.active = True
            anchor_lvl = self.grid_levels[self.last_grid_index] if self.last_grid_index is not None else "None"
            return ("LOG", f"Grid Bot Active @ ${current_price}. Anchor Level: {anchor_lvl}", 0, "GridBot")

        # Check for grid crossing
        new_index = None
        for i, level in enumerate(self.grid_levels):
            if current_price >= level:
                new_index = i
        
        if new_index is not None and self.last_grid_index is not None:
            if new_index > self.last_grid_index:
                # Price moved up -> SELL at the levels we crossed
                self.last_grid_index = new_index
                sz = self.investment_amount / self.grids / current_price
                return ("SELL", self.grid_levels[new_index], sz, "GridBot")
            elif new_index < self.last_grid_index:
                # Price moved down -> BUY at the level we crossed
                self.last_grid_index = new_index
                sz = self.investment_amount / self.grids / current_price
                return ("BUY", self.grid_levels[new_index], sz, "GridBot")

        return None

    def update_position(self, side: str, price: float, size: float):
        """Updates internal state based on a confirmed fill."""
        if side.lower() == "buy":
            new_pos = self.current_pos + size
            if new_pos > 0:
                self.avg_price = ((self.avg_price * self.current_pos) + (price * size)) / new_pos
            self.current_pos = new_pos
            logger.info(f"GridBot {self.inst_id} filled BUY: {size} @ {price}. New Pos: {self.current_pos}, Avg: {self.avg_price}")
        
        elif side.lower() == "sell":
            if self.current_pos > 0:
                # Realize PnL against the average price
                pnl_gain = (price - self.avg_price) * size
                self.realized_pnl += pnl_gain
                self.current_pos -= size
                logger.info(f"GridBot {self.inst_id} filled SELL: {size} @ {price}. Realized Gain: {pnl_gain}. Total Realized: {self.realized_pnl}")
            else:
                self.current_pos -= size

    def calculate_pnl(self, current_price: float) -> float:
        """Calculates total PnL (Realized + Unrealized)."""
        unrealized = self.current_pos * (current_price - self.avg_price) if self.current_pos != 0 else 0
        return self.realized_pnl + unrealized

class StrategyManager:
    """Orchestrates multiple active strategy instances."""
    def __init__(self):
        self.active_bots = {} # {bot_id: bot_instance}
        self.total_pnl = 0.0

    @staticmethod
    def calculate_ema(data: list, period: int) -> list:
        """Computes the Exponential Moving Average (EMA) for a given series."""
        if len(data) < period:
            return [None] * len(data)
        
        ema = [None] * len(data)
        multiplier = 2 / (period + 1)
        
        # Start with simple SMA for the first valid point
        initial_sma = sum(data[:period]) / period
        ema[period - 1] = initial_sma
        
        for i in range(period, len(data)):
            ema[i] = (data[i] - ema[i - 1]) * multiplier + ema[i - 1]
            
        return ema

    @staticmethod
    def calculate_rsi(data: list, period: int = 14) -> list:
        """Computes the Relative Strength Index (RSI) using Wilder's smoothing."""
        if len(data) <= period:
            return [None] * len(data)
            
        rsi = [None] * len(data)
        deltas = [data[i] - data[i - 1] for i in range(1, len(data))]
        
        gains = [d if d > 0 else 0 for d in deltas]
        losses = [-d if d < 0 else 0 for d in deltas]
        
        avg_gain = sum(gains[:period]) / period
        avg_loss = sum(losses[:period]) / period
        
        if avg_loss == 0:
            rsi[period] = 100
        else:
            rs = avg_gain / avg_loss
            rsi[period] = 100 - (100 / (1 + rs))
            
        for i in range(period + 1, len(data)):
            # Wilder's Smoothing
            avg_gain = (avg_gain * (period - 1) + gains[i - 1]) / period
            avg_loss = (avg_loss * (period - 1) + losses[i - 1]) / period
            
            if avg_loss == 0:
                rsi[i] = 100
            else:
                rs = avg_gain / avg_loss
                rsi[i] = 100 - (100 / (1 + rs))
                
        return rsi

    def start_grid_bot(self, inst_id: str, lower: float, upper: float, grids: int, investment: float, grid_type: str = "arithmetic"):
        bot_id = f"grid_{inst_id}_{int(time.time())}"
        bot = GridStrategyEngine(inst_id, lower, upper, grids, investment, grid_type)
        self.active_bots[bot_id] = bot
        return bot_id

    def start_dca_bot(self, inst_id: str, base_amount: float, drop_pct: float):
        bot_id = f"dca_{inst_id}_{int(time.time())}"
        bot = DCAStrategyEngine(inst_id, base_amount, drop_pct)
        self.active_bots[bot_id] = bot
        return bot_id

    def update_bot_fill(self, bot_id, side, price, size):
        """Passes a fill event to a specific bot."""
        if bot_id in self.active_bots:
            self.active_bots[bot_id].update_position(side, price, size)

    def get_total_session_pnl(self, current_prices):
        """
        Aggregates PnL across all active bots.
        current_prices: {inst_id: price}
        """
        total = 0.0
        for bot in self.active_bots.values():
            price = current_prices.get(bot.inst_id)
            if price is not None:
                total += bot.calculate_pnl(price)
            else:
                # If no current price, just add realized
                total += bot.realized_pnl
        return total

    def stop_all(self):
        count = len(self.active_bots)
        self.active_bots.clear()
        return count

    def get_status_summary(self):
        bot_details = []
        for bid, bot in self.active_bots.items():
            type_str = "GRID" if isinstance(bot, GridStrategyEngine) else "DCA"
            bot_details.append(f"{type_str} | {bot.inst_id} | Pos: {bot.current_pos:.4f}")
            
        return {
            "count": len(self.active_bots),
            "status": "ACTIVE" if self.active_bots else "IDLE",
            "details": "\n".join(bot_details) if bot_details else "No active bots"
        }

class DCAStrategyEngine:
    def __init__(self, inst_id: str, base_order_size: float, drop_trigger_pct: float):
        self.inst_id = inst_id
        self.base_order_size = base_order_size # In USD
        self.drop_trigger_pct = drop_trigger_pct
        self.last_purchase_price = None
        self.active = False
        self.realized_pnl = 0.0
        self.current_pos = 0.0
        self.avg_price = 0.0

    def process_tick(self, current_price: float):
        """Checks if the price has dropped enough from the last buy to trigger a DCA order."""
        if not self.active:
            self.active = True
            self.last_purchase_price = current_price
            sz = self.base_order_size / current_price
            return ("BUY", current_price, sz, "DCABot")

        drop_pct = ((self.last_purchase_price - current_price) / self.last_purchase_price) * 100
        if drop_pct >= self.drop_trigger_pct:
            logger.info(f"DCA Triggered! Price dropped by {drop_pct:.2f}%")
            self.last_purchase_price = current_price
            sz = self.base_order_size / current_price
            return ("BUY", current_price, sz, "DCABot")

        return None

    def update_position(self, side: str, price: float, size: float):
        if side.lower() == "buy":
            new_pos = self.current_pos + size
            if new_pos > 0:
                self.avg_price = ((self.avg_price * self.current_pos) + (price * size)) / new_pos
            self.current_pos = new_pos
            logger.info(f"DCABot {self.inst_id} filled BUY: {size} @ {price}. New Pos: {self.current_pos}, Avg: {self.avg_price}")
        elif side.lower() == "sell":
            # DCA usually only buys, but we allow sell for manual intervention/TP
            if self.current_pos > 0:
                pnl_gain = (price - self.avg_price) * size
                self.realized_pnl += pnl_gain
                self.current_pos -= size
                logger.info(f"DCABot {self.inst_id} filled SELL: {size} @ {price}. Realized Gain: {pnl_gain}")
            else:
                self.current_pos -= size

    def calculate_pnl(self, current_price: float) -> float:
        unrealized = self.current_pos * (current_price - self.avg_price) if self.current_pos != 0 else 0
        return self.realized_pnl + unrealized

File: CPP/StrategyEngine.cpp
#include "StrategyEngine.hpp"
#include <cmath>
#include <algorithm>
#include <numeric>
#include <iostream>
#include <ctime>

// --------------------------------------------------------------------------------
// 1. OKX Grid Validator
// --------------------------------------------------------------------------------
OKXGridValidator::OKXGridValidator(double min_order_val, int min_grids, int max_grids)
    : min_order_val(min_order_val), min_grids(min_grids), max_grids(max_grids) {}

std::pair<bool, std::string> OKXGridValidator::validate_setup(double lower_price, double upper_price,
                                                               int grid_count, double total_investment,
                                                               double current_market_price) {
    if (lower_price >= upper_price) {
        return {false, "Error: Lower price must be strictly less than upper price."};
    }
    if (current_market_price < lower_price || current_market_price > upper_price) {
        return {false, "Error: Price is outside grid range."};
    }
    if (grid_count < min_grids || grid_count > max_grids) {
        return {false, "Error: Grid count must be between " + std::to_string(min_grids) + " and " + std::to_string(max_grids)};
    }
    double investment_per_grid = total_investment / grid_count;
    if (investment_per_grid < min_order_val) {
        return {false, "Error: Investment per grid is below OKX minimum."};
    }
    return {true, "Validation Passed"};
}

// --------------------------------------------------------------------------------
// 3. Grid Strategy Engine
// --------------------------------------------------------------------------------
GridStrategyEngine::GridStrategyEngine(const std::string& inst_id, double lower_bound, double upper_bound,
                                       int grids, double investment_amount, const std::string& grid_type)
    : inst_id(inst_id), lower_bound(lower_bound), upper_bound(upper_bound),
      grids(grids), investment_amount(investment_amount), grid_type(grid_type) {
    _initialize_grid();
}

void GridStrategyEngine::_initialize_grid() {
    if (grids <= 1) return;

    if (grid_type == "geometric") {
        double ratio = std::pow(upper_bound / lower_bound, 1.0 / (grids - 1));
        for (int i = 0; i < grids; ++i) {
            grid_levels.push_back(lower_bound * std::pow(ratio, i));
        }
    } else {
        double step = (upper_bound - lower_bound) / (grids - 1);
        for (int i = 0; i < grids; ++i) {
            grid_levels.push_back(lower_bound + (i * step));
        }
    }
    std::sort(grid_levels.begin(), grid_levels.end());
}

std::optional<TradingSignal> GridStrategyEngine::process_tick(double current_price) {
    if (!active) {
        for (size_t i = 0; i < grid_levels.size(); ++i) {
            if (current_price >= grid_levels[i]) {
                last_grid_index = static_cast<int>(i);
            }
        }
        active = true;
        std::string anchor = last_grid_index ? std::to_string(grid_levels[*last_grid_index]) : "None";
        return TradingSignal{"LOG", current_price, 0.0, "GridBot: Active, Anchor=" + anchor};
    }

    std::optional<int> new_index;
    for (size_t i = 0; i < grid_levels.size(); ++i) {
        if (current_price >= grid_levels[i]) {
            new_index = static_cast<int>(i);
        }
    }

    if (new_index && last_grid_index) {
        if (*new_index > *last_grid_index) {
            last_grid_index = new_index;
            double sz = (investment_amount / grids) / current_price;
            return TradingSignal{"SELL", grid_levels[*new_index], sz, "GridBot"};
        } else if (*new_index < *last_grid_index) {
            last_grid_index = new_index;
            double sz = (investment_amount / grids) / current_price;
            return TradingSignal{"BUY", grid_levels[*new_index], sz, "GridBot"};
        }
    }
    return std::nullopt;
}

void GridStrategyEngine::update_position(const std::string& side, double price, double size) {
    if (side == "BUY" || side == "buy") {
        double new_pos = current_pos + size;
        if (new_pos > 0) {
            avg_price = ((avg_price * current_pos) + (price * size)) / new_pos;
        }
        current_pos = new_pos;
    } else if (side == "SELL" || side == "sell") {
        if (current_pos > 0) {
            realized_pnl += (price - avg_price) * std::min(current_pos, size);
        }
        current_pos -= size;
    }
}

double GridStrategyEngine::calculate_pnl(double current_price) const {
    double unrealized = (current_pos != 0) ? current_pos * (current_price - avg_price) : 0.0;
    return realized_pnl + unrealized;
}

// --------------------------------------------------------------------------------
// 4. DCA Strategy Engine
// --------------------------------------------------------------------------------
DCAStrategyEngine::DCAStrategyEngine(const std::string& inst_id, double base_order_size, double drop_trigger_pct)
    : inst_id(inst_id), base_order_size(base_order_size), drop_trigger_pct(drop_trigger_pct) {}

std::optional<TradingSignal> DCAStrategyEngine::process_tick(double current_price) {
    if (!active) {
        active = true;
        last_purchase_price = current_price;
        return TradingSignal{"BUY", current_price, base_order_size / current_price, "DCABot"};
    }

    if (last_purchase_price) {
        double drop = ((*last_purchase_price - current_price) / *last_purchase_price) * 100.0;
        if (drop >= drop_trigger_pct) {
            last_purchase_price = current_price;
            return TradingSignal{"BUY", current_price, base_order_size / current_price, "DCABot"};
        }
    }
    return std::nullopt;
}

void DCAStrategyEngine::update_position(const std::string& side, double price, double size) {
    if (side == "BUY" || side == "buy") {
        double new_pos = current_pos + size;
        if (new_pos > 0) {
            avg_price = ((avg_price * current_pos) + (price * size)) / new_pos;
        }
        current_pos = new_pos;
    } else if (side == "SELL" || side == "sell") {
        if (current_pos > 0) {
            realized_pnl += (price - avg_price) * std::min(current_pos, size);
        }
        current_pos -= size;
    }
}

double DCAStrategyEngine::calculate_pnl(double current_price) const {
    double unrealized = (current_pos != 0) ? current_pos * (current_price - avg_price) : 0.0;
    return realized_pnl + unrealized;
}

// --------------------------------------------------------------------------------
// 5. Strategy Manager
// --------------------------------------------------------------------------------
std::string StrategyManager::start_grid_bot(const std::string& inst_id, double lower, double upper,
                                            int grids, double investment, const std::string& grid_type) {
    std::string bot_id = "grid_" + inst_id + "_" + std::to_string(std::time(nullptr));
    active_bots[bot_id] = std::make_unique<GridStrategyEngine>(inst_id, lower, upper, grids, investment, grid_type);
    return bot_id;
}

std::string StrategyManager::start_dca_bot(const std::string& inst_id, double base_amount, double drop_pct) {
    std::string bot_id = "dca_" + inst_id + "_" + std::to_string(std::time(nullptr));
    active_bots[bot_id] = std::make_unique<DCAStrategyEngine>(inst_id, base_amount, drop_pct);
    return bot_id;
}

void StrategyManager::update_bot_fill(const std::string& bot_id, const std::string& side, double price, double size) {
    auto it = active_bots.find(bot_id);
    if (it != active_bots.end()) {
        it->second->update_position(side, price, size);
    }
}

double StrategyManager::get_total_session_pnl(const std::map<std::string, double>& current_prices) {
    double total = 0.0;
    for (auto const& [id, bot] : active_bots) {
        auto it = current_prices.find(bot->get_inst_id());
        if (it != current_prices.end()) {
            total += bot->calculate_pnl(it->second);
        } else {
            total += bot->get_realized_pnl();
        }
    }
    return total;
}

int StrategyManager::stop_all() {
    int count = static_cast<int>(active_bots.size());
    active_bots.clear();
    return count;
}

// --------------------------------------------------------------------------------
// Technical Indicators
// --------------------------------------------------------------------------------
std::vector<std::optional<double>> StrategyManager::calculate_ema(const std::vector<double>& data, int period) {
    std::vector<std::optional<double>> ema(data.size(), std::nullopt);
    if (data.size() < static_cast<size_t>(period)) return ema;

    double multiplier = 2.0 / (period + 1);
    double initial_sma = std::accumulate(data.begin(), data.begin() + period, 0.0) / period;
    ema[period - 1] = initial_sma;

    for (size_t i = period; i < data.size(); ++i) {
        ema[i] = (data[i] - *ema[i - 1]) * multiplier + *ema[i - 1];
    }
    return ema;
}

std::vector<std::optional<double>> StrategyManager::calculate_rsi(const std::vector<double>& data, int period) {
    std::vector<std::optional<double>> rsi(data.size(), std::nullopt);
    if (data.size() <= static_cast<size_t>(period)) return rsi;

    std::vector<double> deltas;
    for (size_t i = 1; i < data.size(); ++i) deltas.push_back(data[i] - data[i - 1]);

    double avg_gain = 0, avg_loss = 0;
    for (int i = 0; i < period; ++i) {
        if (deltas[i] > 0) avg_gain += deltas[i];
        else avg_loss -= deltas[i];
    }
    avg_gain /= period;
    avg_loss /= period;

    if (avg_loss == 0) rsi[period] = 100.0;
    else {
        double rs = avg_gain / avg_loss;
        rsi[period] = 100.0 - (100.0 / (1.0 + rs));
    }

    for (size_t i = period + 1; i < data.size(); ++i) {
        double gain = (deltas[i - 1] > 0) ? deltas[i - 1] : 0.0;
        double loss = (deltas[i - 1] < 0) ? -deltas[i - 1] : 0.0;
        avg_gain = (avg_gain * (period - 1) + gain) / period;
        avg_loss = (avg_loss * (period - 1) + loss) / period;

        if (avg_loss == 0) rsi[i] = 100.0;
        else {
            double rs = avg_gain / avg_loss;
            rsi[i] = 100.0 - (100.0 / (1.0 + rs));
        }
    }
    return rsi;
}
